torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -847,9 +849,10 @@ ProcessGroupNCCL::CUDAEventCache& ProcessGroupNCCL::CUDAEventCache::get(
         it,
         std::piecewise_construct,
         std::forward_as_tuple(device),
-        std::forward_as_tuple());
+        std::forward_as_tuple(
+            std::make_shared<ProcessGroupNCCL::CUDAEventCache>()));
   }
-  return it->second;
+  return *(it->second);
```

Comment:
This function may need to return a `std::shared_ptr<ProcessGroupNCCL::CUDAEventCache>>` than a reference. That is, if we start to use `std::shared_ptr` we may probably need to use it all the way. (That is, the original `return it->second` can remain the same.)

Reply:
Is this really needed? I mean after calling `get`, the callsite will do dereference anyway.

Reply:
I can't comment on whether it is really needed, but seems a general practice to not mix the use of shared_ptr with reference.